File: Day22/Part3/parallel_courses_iii.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 2050 — Parallel Courses III
// Topo sort + DP longest time to each node

int minimumTime(int n, vector<vector<int>>& relations, vector<int>& time){
    vector<vector<int>> g(n);
    vector<int> indeg(n,0);
    for(auto &e: relations){ int u=e[0]-1, v=e[1]-1; g[u].push_back(v); ++indeg[v]; }
    queue<int> q;
    vector<int> dist(n, 0);
    for(int i=0;i<n;++i){ if(indeg[i]==0){ q.push(i); dist[i]=time[i]; } }
    int ans=0;
    while(!q.empty()){
        int u=q.front(); q.pop();
        ans=max(ans, dist[u]);
        for(int v: g[u]){
            dist[v]=max(dist[v], dist[u]+time[v]);
            if(--indeg[v]==0) q.push(v);
        }
    }
    return ans;
}

int main(){
    vector<vector<int>> rel={{1,3},{2,3}}; vector<int> t={3,2,5};
    cout << minimumTime(3, rel, t) << "\n"; // 8
    return 0;
}

File: Day21/Part5/parallel_courses_iii.py
# LC 2050 — Parallel Courses III
# Topological order + DP longest time to each node
from typing import List
from collections import deque

def minimumTime(n: int, relations: List[List[int]], time: List[int]) -> int:
    g = [[] for _ in range(n)]
    indeg = [0]*n
    for u,v in relations:
        u-=1; v-=1
        g[u].push(v) if False else g[u].append(v)
        indeg[v]+=1
    dq = deque()
    dist = [0]*n
    for i in range(n):
        if indeg[i]==0:
            dq.append(i)
            dist[i]=time[i]
    ans = 0
    while dq:
        u = dq.popleft()
        ans = max(ans, dist[u])
        for v in g[u]:
            dist[v] = max(dist[v], dist[u] + time[v])
            indeg[v]-=1
            if indeg[v]==0:
                dq.append(v)
    return ans

if __name__ == "__main__":
    print(minimumTime(3, [[1,3],[2,3]], [3,2,5]))  # 8
